### PR TITLE
Add option for running pylint without __init__ file

### PR DESCRIPTION
**Pull request by mzfr**
Fix for #352 
I have added option for running the Lint-all option. 
As I have mentioned in my comment [here](https://github.com/PyCQA/pylint/issues/352#issuecomment-397196973) we'll have to pass the lint-all option to `expand_module` function in [/pylint/utils.py L#1075](https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L1075)
Also See [Comment](https://github.com/PyCQA/pylint/issues/352#issuecomment-397198813) by @PCManticore

### CODE REVIEW

**Comment by mzfr** [comment]
@PCManticore I have added the options as suggestd by you but I am still not getting any where.
What am I doing wrong ?

File: pylint/lint.py [hunk 1]
```diff
@@ -922,8 +929,9 @@ def _do_check(self, files_or_modules):
     def expand_files(self, modules):
         """get modules and errors from a list of modules and handle errors
         """
+        print("running pylint")
         result, errors = utils.expand_modules(modules, self.config.black_list,
-                                              self.config.black_list_re)
+                                              self.config.black_list_re, self.lint_all)
```
Comment (PCManticore) [inline pylint/lint.py]: This is `self.config.lint_all`, not `self.lint_all`.

**Comment by mzfr** [comment]
Will I have to define the `lint_all` variable in the config.py file ?

File: pylint/lint.py (hunk 1 above)
Comment (mzfr) [reply pylint/lint.py]: Will I have to define the `lint_all` variable in the config.py file ?

**Comment by mzfr** [comment]
@PCManticore I have changed it but now I am getting this error:
```traceback
Traceback (most recent call last):
  File "/home/mzfr/dev/Pylint/venv/bin/pylint", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/mzfr/dev/Pylint/pylint/bin/pylint", line 9, in <module>
    load_entry_point('pylint==2.0.0', 'console_scripts', 'pylint')()
  File "/home/mzfr/dev/Pylint/pylint/pylint/__init__.py", line 18, in run_pylint
    Run(sys.argv[1:])
  File "/home/mzfr/dev/Pylint/pylint/pylint/lint.py", line 1366, in __init__
    linter.check(args)
  File "/home/mzfr/dev/Pylint/pylint/pylint/lint.py", line 786, in check
    self._do_check(files_or_modules)
  File "/home/mzfr/dev/Pylint/pylint/pylint/lint.py", line 901, in _do_check
    for descr in self.expand_files(files_or_modules):
  File "/home/mzfr/dev/Pylint/pylint/pylint/lint.py", line 933, in expand_files
    self.config.black_list_re, self.config.lint_all)
  File "/home/mzfr/dev/Pylint/pylint/pylint/utils.py", line 956, in expand_modules
    modpath = _modpath_from_file(subfilepath, is_namespace)
  File "/home/mzfr/dev/Pylint/pylint/pylint/utils.py", line 892, in _modpath_from_file
    return modutils.modpath_from_file_with_callback(filename, is_package_cb=_is_package_cb)
  File "/home/mzfr/dev/Pylint/astroid/astroid/modutils.py", line 310, in modpath_from_file_with_callback
    filename, ', \n'.join(sys.path)))
ImportError: Unable to find module for /home/mzfr/dev/Addon-check/addons/plugin.video.twitch/resources/lib/service.py in /home/mzfr/dev/Addon-check/addons/plugin.video.twitch,

```

I am using the command: `pylint /home/mzfr/dev/Addon-check/addons/plugin.video.twitch/ --lint-all y`

Why is it not able to find any file within the same folder ?

**Comment by mzfr** [comment]
On [line 340 in modutils.py](https://github.com/PyCQA/astroid/blob/master/astroid/modutils.py#L340) why are we doing:
```
            if is_package_cb(path, modpath[:-1]):
                return modpath
```

**Comment by mzfr** [comment]
All the test in pylint is failing because the `get_module_file` currently takes only 3 argument and I have made changes to take 4 arguments.

**Comment by mzfr** [comment]
@PCManticore I have done it :smile: Now pylint will work both way i.e with or without the presence of `__init__.py` file.

There wasn't any changes required in astroid so I am closing the [pull #560](https://github.com/PyCQA/astroid/pull/560).

**Comment by mzfr** [comment]
@PCManticore Thanks :slightly_smiling_face: 
Point 2 and 3 are easy to do but I have question related to point 1.

1) will I have to make a new function for this test or will I have to make changes to an existing function. If latter then Can you specify which one ?
If a new function then do you want any specific name for the function ?

2) > Basically you can have a directory in regr_data and from test_self we trigger a pylint run with and without the flag, asserting that the correct output for each branch is emitted.

This is what I am able to understand by this => I will make a new directory in `regrtest_data` name `bad_module` which will not have a `__init__.py` file in it and will only have a random `.py` file and then using the t`est_self.py` I will run pylint on this `bad_module` folder and check for the output in both the cases. Let me know if am wrong.

**Comment by PCManticore** [comment]
@mzfr A new test should be fine, any name that suggest that it does should okay. Yep, you got it right, you can also reuse this PR for the tests.

**Comment by mzfr** [comment]
This is the test I have added to test_self.py

```
 def test_pylint_with_or_without_init_file(self):
        module = join(HERE, 'regrtest_data', 'bad_module')

        expected_output = textwrap.dedent("""
            ************* Module not_init
            C:  1, 0: Missing module docstring (missing-docstring)
            C:  2, 0: Constant name "string" doesn't conform to UPPER_CASE naming style (invalid-name)
            W:  1, 0: Unused import pylint (unused-import)
            """)
        self._test_output([module, "--lint-all y"], expected_output=expected_output)

        expected_output = textwrap.dedent("""
            ************
            F0010: 1: error while code parsing: Unable to load file
            """)
        self._test_output([module], expected_output=expected_output)

```

This is what file structure looks like for regrtest_data/bad_modules

```
├── bad_module
│   └── not_init.py
```

The content of the file `not_init.py` is:

```
import pylint
string = "This line is going to be long enough so pylint mark it with Line too long tag thing-y"
```

If I run pylint with command line then this is what I get:

* `pylint /home/mzfr/dev/Pylint/pylint/pylint/test/regrtest_data/bad_module/`
```
*************
F:  1, 0: error while code parsing: Unable to load file /home/mzfr/dev/Pylint/pylint/pylint/test/regrtest_data/bad_module/__init__.py:
[Errno 2] No such file or directory: '/home/mzfr/dev/Pylint/pylint/pylint/test/regrtest_data/bad_module/__init__.py' (parse-error)
```

* `pylint /home/mzfr/dev/Pylint/pylint/pylint/test/regrtest_data/bad_module/ --lint-all y`
```
************* Module not_init
C:  1, 0: Missing module docstring (missing-docstring)
C:  2, 0: Constant name "string" doesn't conform to UPPER_CASE naming style (invalid-name)
W:  1, 0: Unused import pylint (unused-import)
```

I am getting an assertion error. 
Also in first example it says `error while code parsing` and then there is a path to the file how can I add that path ?

**Comment by mzfr** [comment]
Hi, I am sorry to bother you again but I am having problem with tests. I am
not sure whether you have seen my comment
<https://github.com/PyCQA/pylint/pull/2190#issuecomment-399056237> or not
but if you haven't then I would request you to please have a look. In my
comment I have mentioned the problem I am having.
I am pushing things because My GSoC proposal includes implementing the
pylint integration into addon-checker-tool
<https://github.com/xbmc/addon-check>
Hope you'll understand.

Thanks

**Comment by PCManticore** [comment]
@mzfr To be honest I'm not entirely sure why this improvement was tied to your GSoc application. While your application is more rigid in terms of time, for me pylint is a tiny part of my life, I have other responsabilities other than monitoring the PRs and the issue tracker.

Regarding the test issues, I took a look on your branch:

- you should be set if you use `--lint-all=y` instead of `--lint-all y` in the tests. I think the latter becomes two options instead of one resulting in your test failure.

- there seems to be a bug with `lint_all` and replacing it for `is_namespace` in `expand_modules`, which reproduces when you run `pylint` over a single file. If you try to run `pylint` just on that file that you're using in the test, what's going to happen is that we won't add any results in the `result` list, since we're explicitly ignoring those for namespace packages [here](https://github.com/PyCQA/pylint/blob/3cf841a6a22360b28938b7e18116066a18f59532/pylint/utils.py#L1061). The reason for that is that when we have namespace packages, we already know they are directories so we're only interested in the inner files, while in this case we'd want the same file as well if we're linting a file.

- regarding the last point, you should also test running on a package without `__init__`, not just over a file, as that's what we're mostly interested in.

- regarding adding the file path when it errors with `error while code parsing`, you can insert the file path into your expected output, which can be fetched with `join(HERE, <the directory>, <the filepath that errors>)` or something of sorts.

Hope this helps!

File: pylint/test/test_self.py [hunk 2]
```diff
@@ -478,3 +478,20 @@ def test_parseable_file_path(self):
...
+    def test_pylint_with_or_without_init_file(self):
+        module = join(HERE, 'regrtest_data', 'bad_module/not_init.py')
+
+        expected_output = textwrap.dedent("""
+            ************* Module not_init
+            C:  1, 0: Missing module docstring (missing-docstring)
+            C:  2, 0: Constant name "string" doesn't conform to UPPER_CASE naming style (invalid-name)
+            W:  1, 0: Unused import pylint (unused-import)
+            """)
+        self._test_output([module, "--lint-all y"], expected_output=expected_output)
```
Comment (Rechi) [inline pylint/test/test_self.py]: `"--lint-all=y"`

**Comment by PCManticore** [comment]
Closing as this still needs some work before being ready for inclusion.

**Pull request closed without merging**